packages/react-server/src/ReactDOMServerFormatConfig.js

```diff
@@ -7,17 +7,367 @@
+import {
+  writeChunk,
+  convertStringToBuffer,
+} from 'react-server/src/ReactServerStreamConfig';
+
+import invariant from 'shared/invariant';
+
+// Per response,
+export type ResponseState = {
+  sentCompleteSegmentFunction: boolean,
+  sentCompleteBoundaryFunction: boolean,
+  sentClientRenderFunction: boolean,
+};
+
+// Allows us to keep track of what we've already written so we can refer back to it.
+export function createResponseState(): ResponseState {
+  return {
+    sentCompleteSegmentFunction: false,
+    sentCompleteBoundaryFunction: false,
+    sentClientRenderFunction: false,
+  };
+}
+
+// This object is used to lazily reuse the ID of the first generated node, or assign one.
+// This is very specific to DOM where we can't assign an ID to.
+export type SuspenseBoundaryID = {
+  id: null | string,
+};
+
+export function createSuspenseBoundaryID(
+  responseState: ResponseState,
+): SuspenseBoundaryID {
+  return {id: null};
+}
+
+function encodeHTMLIDAttribute(value: string): string {
+  // TODO: This needs to be encoded for security purposes.
+  return value;
+}
+
+function encodeHTMLTextNode(text: string): string {
+  // TOOD: This needs to be encoded for security purposes.
+  return text;
+}
+
+export function pushTextInstance(
+  target: Array<Uint8Array>,
+  text: string,
+): void {
+  target.push(convertStringToBuffer(encodeHTMLTextNode(text)));
+}
+
+const startTag1 = convertStringToBuffer('<');
+const startTag2 = convertStringToBuffer('>');
+
+export function pushStartInstance(
+  target: Array<Uint8Array>,
+  type: string,
+  props: Object,
+): void {
+  // TODO: Figure out if it's self closing and everything else.
+  target.push(startTag1, convertStringToBuffer(type), startTag2);
+}
+
+const endTag1 = convertStringToBuffer('</');
+const endTag2 = convertStringToBuffer('>');
+
+export function pushEndInstance(
+  target: Array<Uint8Array>,
+  type: string,
+  props: Object,
+): void {
+  // TODO: Figure out if it was self closing.
+  target.push(endTag1, convertStringToBuffer(type), endTag2);
+}
+
+// Structural Nodes
+
+// A placeholder is a node inside a hidden partial tree that can be filled in later, but before
+// display. It's never visible to users.
+const placeholder1 = convertStringToBuffer('<span id="');
+const placeholder2 = convertStringToBuffer('P:');
+const placeholder3 = convertStringToBuffer('"></span>');
+export function writePlaceholder(
+  destination: Destination,
+  id: number,
+): boolean {
+  // TODO: This needs to be contextually aware and switch tag since not all parents allow for spans like
+  // <select> or <tbody>. E.g. suspending a component that renders a table row.
+  writeChunk(destination, placeholder1);
+  // TODO: Use the identifierPrefix option to make the prefix configurable.
+  writeChunk(destination, placeholder2);
+  const formattedID = convertStringToBuffer(id.toString(16));
+  writeChunk(destination, formattedID);
+  return writeChunk(destination, placeholder3);
```

Comment:
I'm curious if it's an intentional part of the design that you write a lot of really small chunks (separate function calls for placeholder 1, 2, etc) ? Are there performance implications of enqueueing all of these individually or would you expect it to be negligible.

Reply:
It depends a bit on the underlying API. If it's suboptimal, like Node.js certainly is, then yea there can be quite a bit of overhead in the additional calls. There's opportunities to lower it but there will always be some.

However, when the thing you're writing is a static chunk like the ones I hoisted, you save quite a bit compared to trying to merge them together. If it's dynamic strings then you might be better off concatenating the strings. If the static chunks are small, then it might also be worth while tacking them onto a dynamic string.

It's a tradeoff we can play with.

I intentionally start with the most esoteric form that people aren't aware of to make everyone aware of the option to e.g. use a hoisted shared array buffer as a technique.

Also, converting *everything* to buffers using convertStringToBuffer is suboptimal because certain sinks can use the string as is and emit it directly to the underlying buffer. However, some APIs (like web streams) don't support this yet.

I plan on backtracking this to mix strings and buffers if both are available (like on Node). Mostly buffers for static stuff so that it can safely skip the encoding part and mostly strings for dynamic stuff so that the implementation can choose if it can go direct or needs to convert it.

Again, buffers is the least idiomatic one so I start with that to preserve the option to use them.